pinot-transport/src/main/java/com/linkedin/pinot/routing/builder/LargeClusterRoutingTableBuilder.java

```diff
@@ -0,0 +1,414 @@
+
+        segmentToReplicaSetQueue.add(new ImmutablePair<>(segment, instancesForThisSegment));
+      }
+
+      // Create the routing table from the segment -> instances priority queue
+      Map<String, Set<String>> instanceToSegmentSetMap = new HashMap<>();
+      int[] replicas = new int[10];
+      while(!segmentToReplicaSetQueue.isEmpty()) {
+        Pair<String, Set<String>> segmentAndReplicaSet = segmentToReplicaSetQueue.poll();
+        String segment = segmentAndReplicaSet.getKey();
+        Set<String> replicaSet = segmentAndReplicaSet.getValue();
+        replicas[replicaSet.size() - 1]++;
+
+        String instance = pickWeightedRandomReplica(replicaSet, instanceToSegmentSetMap);
+        if (instance != null) {
+          Set<String> segmentsAssignedToInstance = instanceToSegmentSetMap.get(instance);
+
+          if (segmentsAssignedToInstance == null) {
+            segmentsAssignedToInstance = new HashSet<>();
+            instanceToSegmentSetMap.put(instance, segmentsAssignedToInstance);
+          }
+
+          segmentsAssignedToInstance.add(segment);
+        }
+      }
+
+      // Compute disparity metric
+      int maxSegmentCountPerServer = 0;
+      int minSegmentCountPerServer = Integer.MAX_VALUE;
+      for (Set<String> segmentsForServer : instanceToSegmentSetMap.values()) {
+        int serverSegmentCount = segmentsForServer.size();
+
+        if (serverSegmentCount < minSegmentCountPerServer) {
+          minSegmentCountPerServer = serverSegmentCount;
+        }
+
+        if (maxSegmentCountPerServer < serverSegmentCount) {
+          maxSegmentCountPerServer = serverSegmentCount;
+        }
+      }
+
+      int disparity = maxSegmentCountPerServer - minSegmentCountPerServer;
+
+      // Add the routing table to the queue
+      routingTablesAndDisparityQueue.add(new ImmutablePair<>(instanceToSegmentSetMap, disparity));
+    }
+
+    // Pick the top 80% of the routing tables with the smallest disparity (difference between max and min segment count
+    // per server)
+    List<ServerToSegmentSetMap> routingTables = new ArrayList<>(instanceToSegmentMap.size());
+    int routingTablePickCount = (int)(routingTablesAndDisparityQueue.size() * 0.8) + 1;
+    for (int i = 0; i < routingTablePickCount; i++) {
+      routingTables.add(new ServerToSegmentSetMap(routingTablesAndDisparityQueue.poll().getKey()));
+    }
+
+    long endTime = System.currentTimeMillis();
+
+    // TODO jfim: Remove this testing code
+    long totalTime = (endTime - startTime);
+    int routingTableCount = routingTables.size();
+    //System.out.println("totalTime = " + totalTime);
+    //System.out.println("routingTableCount = " + routingTableCount);
+
+    return routingTables;
+  }
+
+  public static void main(String[] args) throws Exception {
+    File[] coloDirs = new File("/Users/jfim/work/ev/dec-9").listFiles();
+    Stream<File> jsonFiles = Arrays.stream(coloDirs).flatMap(coloDir -> Arrays.stream(coloDir.listFiles()));
```

Comment:
This will get done as part of the unit tests. Right now there's not even a validation that the routing table hits all the segments in the EV, for example (even though it does).

There's still a fair bit of cleanup left before this is ready to be merged in anyway. :)